reprclust/cluster_methods.py

```diff
@@ -0,0 +1,225 @@
+# -*- coding: utf-8 -*-
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the module for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Module containing cluster methods to uniform calling
+"""
+
+from joblib import Parallel, delayed
+
+import numpy as np
+
+from scipy.spatial.distance import pdist, hamming
+from scipy.cluster.hierarchy import complete
+from scipy.stats import rankdata
+
+from sklearn.cluster import KMeans
+from sklearn.cluster.hierarchical import _hc_cut, ward_tree
+from sklearn.metrics.cluster import (adjusted_rand_score,
+                                     adjusted_mutual_info_score,
+                                     contingency_matrix)
+from sklearn.mixture import GMM
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.utils.linear_assignment_ import linear_assignment
+
+
+def cut_tree_scipy(Y, k):
+    """ Given the output Y of a hierarchical clustering solution from scipy
+    and a number k, cuts the tree and returns the labels.
+    """
+    children = Y[:, 0:2].astype(int)
+    # convert children to correct values for _hc_cut
+    return _hc_cut(k, children, len(children)+1)
+
+def _predict_knn(self, newdata, k, n_neighbors=1):
+    """Common function to be used to predict clustering solution when method
+    doesn't have built-in prediction method"""
+    if k not in self._clusters:
+        raise ValueError('Run cluster solution of {0} first'.format(k))
+    # if it gets called multiple times, data get rewritten
+    labels = self.get_clusters(k)
+    # train a classifier on this clustering
+    knn = KNeighborsClassifier(n_neighbors=n_neighbors)
+    knn.fit(self.data, labels)
+    # predict
+    predicted_labels = knn.predict(newdata)
+    self._predicted[k] = predicted_labels
+
+
+def _predict(self, newdata, k):
+    """Common function to be used to predict clustering solution when method
+    has built-in prediction method"""
+    self._predicted[k] = self._methods[k].predict(newdata)
+
+
+class ClusterMethod(object):
+    """Wrapper for clustering methods to add predict method and provide
+    solution for different ks
+
+    Attributes
+    ----------
+    clustering_method : callable
+        A function that clusters data.
+
+    Methods
+    -------
+    cluster(*args, **kwargs)
+        Clusters the data. args and kwargs depend on the current method.
+    predict(*args, **kwargs)
+        Predicts clustering on new data based on previous cluster solution.
+        args and kwargs depend on the current method.
+    get_clusters(k)
+        Returns the cluster solution for k clusters.
+    get_predicted_clusters(k)
+        Returns the predicted cluster solution for k clusters.
+    """
+    def __init__(self, clustering_method):
```

Comment:
no need to pass the specific implementation  function here. Derived classes will call necessary function themselves.  Generally you don't want to mask out parameters for the methods of the parent class -- you just want to extend its API (e.g. with new parameters like parameters for the specific clustering implementation)